File: source/peripherals/rotary_encoder.py
from event import Event
from machine import Pin

class RotaryEncoder ():
    def __init__(self, clk_pin, dt_pin):

        self.dt_pin = dt_pin
        self.clk_pin = clk_pin
        self.last_clk_value = clk_pin.value()

    @property
    def direction(self):
        clk_value = self.clk_pin.value()
        dir =  (1 if clk_value == self.dt_pin.value() else -1) if not clk_value == self.last_clk_value else 0 
        self.last_clk_value = clk_value
        return dir

File: firmware/src/inputs/rotary_encoder.cpp
#include <Arduino.h>

#include "rotary_encoder.h"

RotaryEncoder::RotaryEncoder (int dtPin, int clkPin)
{
    this->dt = dtPin;
    this->clk = clkPin;
}

int RotaryEncoder::direction ()
{
    bool clkValue = digitalRead(clk);
    bool dtValue = digitalRead(dt);

    int dir = 0;

    if (clkValue !=  lastClk){
        dir = clkValue && dtValue ? 1 : -1;
        lastClk = clkValue;
    }

    return dir;
}
